adapters/postgres.go

```diff
@@ -479,48 +485,77 @@ func (p *Postgres) BulkInsert(table *Table, objects []map[string]interface{}) er
-	var placeholders string
+	start := time.Now()
+	var placeholdersBuilder strings.Builder
 	var header []string
-	i := 1
 	for name := range table.Columns {
 		header = append(header, name)
-
-		//$1::type, $2::type, $3, etc
-		castClause := ""
-		castType, ok := p.mappingTypeCasts[name]
-		if ok {
-			castClause = "::" + castType
-		}
-		placeholders += "$" + strconv.Itoa(i) + castClause + ","
-
-		i++
 	}
-
-	query := p.insertQuery(table.GetPKFields(), table.Name, strings.Join(header, ","), removeLastComma(placeholders))
-
-	insertStmt, err := wrappedTx.tx.PrepareContext(p.ctx, query)
-	if err != nil {
-		return fmt.Errorf("Error preparing bulk insert statement [%s] table %s statement: %v", query, table.Name, err)
+	maxValues := len(objects) * len(table.Columns)
+	if maxValues > postgresValuesLimit {
+		maxValues = postgresValuesLimit
 	}
-
+	valueArgs := make([]interface{}, 0, maxValues)
+	placeholdersCounter := 1
 	for _, row := range objects {
-		var values []interface{}
-		for _, column := range header {
+		// if number of values exceeds limit, we have to execute insert query on processed rows
+		if len(valueArgs)+len(header) > postgresValuesLimit {
+			err := p.executeInsert(wrappedTx, table, header, placeholdersBuilder, valueArgs)
+			if err != nil {
+				return err
+			}
+			logging.Infof("Inserted %d rows", len(valueArgs)/len(table.Columns))
+			placeholdersBuilder.Reset()
+			placeholdersCounter = 1
+			valueArgs = make([]interface{}, 0, maxValues)
+		}
+		_, err := placeholdersBuilder.WriteString("(")
+		if err != nil {
+			return fmt.Errorf(placeholdersStringBuildErrTemplate, err)
+		}
+		for i, column := range header {
 			value, _ := row[column]
-			values = append(values, value)
+			valueArgs = append(valueArgs, value)
+			castClause := ""
+			castType, ok := p.mappingTypeCasts[column]
+			if ok {
+				castClause = "::" + castType
+			}
+			_, err = placeholdersBuilder.WriteString("$" + strconv.Itoa(placeholdersCounter) + castClause)
+			if err != nil {
+				return fmt.Errorf(placeholdersStringBuildErrTemplate, err)
+			}
+
+			if i < len(header)-1 {
+				_, err = placeholdersBuilder.WriteString(",")
+				if err != nil {
+					return fmt.Errorf(placeholdersStringBuildErrTemplate, err)
+				}
+			}
+			placeholdersCounter++
 		}
-
-		p.queryLogger.LogQueryWithValues(query, values)
-
-		_, err = insertStmt.ExecContext(p.ctx, values...)
+		_, err = placeholdersBuilder.WriteString("),")
 		if err != nil {
-			return fmt.Errorf("Error bulk inserting in %s table with statement: %s values: %v: %v", table.Name, query, values, err)
+			return fmt.Errorf(placeholdersStringBuildErrTemplate, err)
 		}
 	}
-
+	if len(valueArgs) > 0 {
+		err := p.executeInsert(wrappedTx, table, header, placeholdersBuilder, valueArgs)
+		if err != nil {
+			return err
+		}
+	}
+	logging.Infof("Inserted [%d] rows in [%.2f] seconds", len(objects), time.Now().Sub(start).Seconds())
```

Comment:
This log message isn't clear. We should know the context (destination id).
Maybe we could move it to SQL debug logs?

Reply:
I'm not sure if the message should be at SQL logs. SQL logs idea was that we can debug queries. This information may be useful in application logs (to see the time spent to insert data in case something goes wrong or slow). So, I suppose that this log should be in app logs. I think it may be in Debug mode.

Reply:
Also, at the adapter level, we don't know about the destination id. We may pass it as an additional parameter when an adapter is created. Or we may log data from data source configuration (host, schema, database + table name). The second option seems more confusing, so, I suppose that it's better to let the adapter know the destination id adapter belongs to.

Reply:
Moved logging to the storage level to keep destination id